### PR TITLE
migrate/project tasks list

### PR DESCRIPTION
**Pull request by Nightcatclub**
_Any italic text should be deleted from the final Pull Request text, including this line_

# Description

Migrate "Project-tasks-list" to typescript, It supports users to view the Task submitted by students under Tutor account, and it supports users to click the "No.XX" button in the list to view the Assignment submitted by students.

## Type of change

_Please delete options that are not relevant._

- [x] Component Migration

# How Has This Been Tested?

At the beginning my Npm test never passed. But after checking and fixing the code and test environment, it successfully passed the test on Chrom. Still haven't implemented the test on other browsers

## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [ ] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have requested a review from @PerryRose  , @tancnle and @RayGuo-ergou  on the Pull Request

### CODE REVIEW

**Comment by Nightcatclub** [comment]
[project-tasks-list migration](https://trello.com/c/qakOBd84/59-project-tasks-list-migration)

**Comment by tancnle** [comment]
@Nightcatclub There seems to be an issue with one of the committed files. The CI build is failing on the linting step. There is a lot of existing warnings and it perhaps has obfuscated the errors in the log output. Can you run this command locally and fix the error?

```shell
npx eslint src/app/tasks/project-tasks-list/project-tasks-list.component.ts
```

File: src/app/tasks/project-tasks-list/project-tasks-list.component.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        mat-raised-button
+        matTooltip="{{ taskDefinition(item).name }}:{{ item.status_txt() }}"
+        matTooltipPosition="below"
+        matTooltipShowDelay="200"
+        class="col-xs btn task-status {{ statusClass(item.status) }}"
+        (click)="onClickItem(item, 'Task Button List')"
+        style="width: 100%"
+      >
+        <i class="fa fa-eye" *ngIf="item.similar_to_count - item.similar_to_dismissed_count > 0"></i>
+        {{ taskText(item) }}
```
Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.html:14]: If you do decide to display the grade and rating values in the tooltip, you could remove the `taskText` function entirely, and have `item.definition.abbreviation` instead. 

You could then have a function to get the tooltip label, which could push a string to an array for the grade value if it's graded, another string for the rating if it's rated, then join the array by a space and return it.

File: src/app/tasks/project-tasks-list/project-tasks-list.component.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<ul class="project-tasks-list">
+  <div *ngIf="project.tasks.length > 0" class="group-set-tasks clearfix">
+    <li *ngFor="let item of project.tasks" class="col-xs-4 col-lg-3">
```
Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.html:3]: Not an issue, but renaming `item` to `task` might be a better word to use.

Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.html:3]: Not an issue, but renaming `item` to `task` might be a better word to use.

File: src/app/tasks/project-tasks-list/project-tasks-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    return this.ts.statusText;
+  }
+
+  get hideGroupSetName(): boolean {
+    return this.unit.group_sets.length === 0;
+  }
+
+  public groupSetName(id): string {
+    return this.grounds.groupSetName(id, this.unit);
+  }
```
Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.ts:35]: Are any of these being used? I couldn't find anything that was calling them

File: src/app/tasks/project-tasks-list/project-tasks-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  public statusClass(status) {
+    return this.ts.statusClass(status);
+  }
+
+  public taskDefinition(task) {
+    return this.unit.taskDef(task.task_definition_id)
+  }
```
Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.ts:43]: Is this being used anywhere?

File: src/app/tasks/project-tasks-list/project-tasks-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+mifr
+  public onClickItem(data) {
+    this.onSelect.emit(data);
```
Comment (PerryRose) [inline src/app/tasks/project-tasks-list/project-tasks-list.component.ts:46]: This method is emitting the task object when a task button is clicked on, but nothing appears to happen. I can see `onSelect` is being given a `updateBurndownChart` method from inside `project-progress-dashboard`, but that method doesn't exist on the parent component.

**Review by PerryRose — COMMENTED** [review]
Great job with the migration, Renhao. I have included some feedback in the pull request.

I'm getting some text overflow for tasks that have a rating, grade, or both:
image

I wonder whether having the grade and rating values as the tooltip would be a better option? It cleans up the label a lot more, what do you think?
image

Also, if you could include a few more unit tests, that would be great 🙂
Great job!

**Review by RayGuo-ergou — COMMENTED** [review]
Good Job @Nightcatclub , also thanks @PerryRose for the review. I think most of my suggestion Perry already list, just one CSS class suggestion.

File: src/app/tasks/project-tasks-list/project-tasks-list.component.ts (hunk 3 above)
Comment (RayGuo-ergou) [reply src/app/tasks/project-tasks-list/project-tasks-list.component.ts:35]: If you need to use these functions I think the best way is init the variable then assign values in ` ngoninit` function. (You are not using ` ngoninit` function so need to remove that as well if not functional). If the value will change try to find out if the ` ts.statusText` is an observerable so you can subscribe to that observerable object. Also `ts` is not a good way to name a variable should be clear like `taskService`. Same for ` groups` and `grades` should change to ` groupservices` and ` gradeservices`. ` groups ` sounds like an array contain all the group.

**Review by tancnle — COMMENTED** [review]
Thanks, @Nightcatclub 👍🏼  I have added some suggestions for your consideration. Please let me know what you think.

**Comment by maddernd** [comment]
changed PR to upstream.

**Pull request closed without merging**